internal/observability/event/node.go

```diff
@@ -0,0 +1,163 @@
+	}
+	opts := getOpts(opt...)
+	n := Node{
+		FormatterFilter: &cloudevents.FormatterFilter{
+			Source: source,
+			Schema: opts.withSchema,
+			Format: format,
+		},
+	}
+
+	// intentionally not checking of allow and/or deny optional filters were
+	// supplied since having a filter node with no filters is okay.
+
+	if len(opts.withAllow) > 0 {
+		n.allow = make([]*filter, 0, len((opts.withAllow)))
+		for i := range opts.withAllow {
+			f, err := newFilter(opts.withAllow[i])
+			if err != nil {
+				return nil, fmt.Errorf("%s: invalid allow filter '%s': %w", op, opts.withAllow[i], err)
+			}
+			n.allow = append(n.allow, f)
+		}
+	}
+	if len(opts.withDeny) > 0 {
+		n.deny = make([]*filter, 0, len((opts.withDeny)))
+		for i := range opts.withDeny {
+			f, err := newFilter(opts.withDeny[i])
+			if err != nil {
+				return nil, fmt.Errorf("%s: invalid deny filter '%s': %w", op, opts.withDeny[i], err)
+			}
+			n.deny = append(n.deny, f)
+		}
+	}
+	n.Predicate = newPredicate(n.allow, n.deny)
+	return &n, nil
+}
+
+func newPredicate(allow, deny []*filter) func(ctx context.Context, ce interface{}) (bool, error) {
+	return func(ctx context.Context, ce interface{}) (bool, error) {
+		if len(allow) == 0 && len(deny) == 0 {
+			return true, nil
+		}
+		for _, f := range deny {
+			if f.Match(ce) {
+				return false, nil
+			}
+		}
+		switch {
+		case len(allow) > 0:
+			for _, f := range allow {
+				if f.Match(ce) {
+					return true, nil
+				}
+			}
+			return false, nil
+		default:
+			return true, nil
+		}
+	}
+}
+
+var _ eventlogger.Node = &Node{}
+
+type filter struct {
+	raw  string
+	eval *bexpr.Evaluator
+}
+
+// newFilter returns a Filter which can be matched against.
+func newFilter(f string) (*filter, error) {
+	const op = "event.newFilter"
+	if f == "" {
+		return nil, fmt.Errorf("%s: missing filter: %w", op, ErrInvalidParameter)
+	}
+	e, err := bexpr.CreateEvaluator(f, bexpr.WithHookFn(wellKnownTypeFilterHook))
+	if err != nil {
+		return nil, fmt.Errorf("%s: %w", op, err)
+	}
+	return &filter{eval: e, raw: f}, nil
+}
+
+// Match returns if the provided interface matches the filter. If the filter
+// does not match the structure of the object being Matched, false is returned.
+func (f *filter) Match(item interface{}) bool {
+	if f.eval == nil {
+		return true
+	}
+	m, err := f.eval.Evaluate(item)
+	// There isn't a clear way to differentiate between a JSON Pointer which doesn't represent
+	// the structure of the object being Matched and a JSON Pointer which references a field which
+	// is part of a sub structure that is nil in this item. Because of this, any filter which would
+	// result in an error using the underlying library is simply interpreted as not a match.
+	return err == nil && m
+}
+
+// wellKnownTypeFilterHook is passed to bexpr to treat all proto well-known
+// types as the types they wrap for comparison.
+func wellKnownTypeFilterHook(v reflect.Value) reflect.Value {
```

Comment:
I believe one of these functions exists already in the codebase (@talanknight wrote one IIRC). Might be good to see if they can be consolidated.

Reply:
The hook @jefferai referred to is in internal/servers/controller/handlers/filtering.go

Reply:
Yeah, I just cut/pasted it. Where would these sort of general functions live in our packages?

Reply:
under `libs`?

Reply:
I've refactored WellKnownTypeFilterHook(...) into a new filter package.